File: hashwithreplacement.py
MAX = 10

class Node:
    def __init__(self):
        self.name = "-----"
        self.mn = 0
        self.chain = -1

class Telephone:
    def __init__(self):
        self.ht = [Node() for _ in range(MAX)]
        self.count = 0

    def hashfun(self, num):
        return num % MAX

    def insert(self):
        if self.count == MAX - 1:
            print("Hash table full")
            return

        name = input("Enter name: ")
        mn = int(input("Enter telephone number: "))
        s = Node()
        s.name = name
        s.mn = mn
        ind = self.hashfun(mn)

        if self.ht[ind].mn == 0:
            self.ht[ind] = s
            self.count += 1
        elif self.hashfun(self.ht[ind].mn) == ind:
            prev = ind
            while self.ht[ind].mn != 0:
                if self.ht[ind].chain != -1:
                    ind = self.ht[ind].chain
                    prev = ind
                else:
                    while self.ht[ind].mn != 0:
                        ind = (ind + 1) % MAX
                    self.ht[prev].chain = ind
            self.ht[ind] = s
            self.count += 1
        else:
            newIndex = ind
            while self.ht[newIndex].mn != 0:
                newIndex = (newIndex + 1) % MAX

            temp = self.ht[ind]
            self.ht[newIndex] = temp
            self.ht[ind] = s

            tempIndex = self.hashfun(temp.mn)
            while self.ht[tempIndex].chain != ind:
                tempIndex = self.ht[tempIndex].chain
            self.ht[tempIndex].chain = newIndex

            self.ht[ind].chain = -1
            self.count += 1

    def find(self):
        num = int(input("Enter the telephone number to search: "))
        ind = self.hashfun(num)
        cnt = 0

        while ind != -1:
            cnt += 1
            if self.ht[ind].mn == num:
                print("Record Found!")
                print(f"Name: {self.ht[ind].name}\tPhone: {self.ht[ind].mn} found at index {ind} comparison {cnt}")
                return
            ind = self.ht[ind].chain
        print("Record Not Found!")

    def delno(self):
        num = int(input("Enter the telephone number to delete: "))
        ind = self.hashfun(num)
        prev = -1

        while ind != -1:
            if self.ht[ind].mn == num:
                print("Record Deleted!")
                self.ht[ind].mn = 0
                self.ht[ind].name = "-----"
                if prev != -1:
                    self.ht[prev].chain = self.ht[ind].chain
                self.ht[ind].chain = -1
                return
            prev = ind
            ind = self.ht[ind].chain
        print("Record Not Found!")

    def display(self):
        print("Index\tName\t\tPhone\t\tChain")
        for i in range(MAX):
            print(f"{i}\t{self.ht[i].name}\t\t{self.ht[i].mn}\t\t{self.ht[i].chain}")

def main():
    t = Telephone()
    while True:
        print("\nEnter your choice")
        print("1. Insert")
        print("2. Display")
        print("3. Find")
        print("4. Delete")
        ch = input("Choice: ")

        if ch == '1':
            t.insert()
        elif ch == '2':
            t.display()
        elif ch == '3':
            t.find()
        elif ch == '4':
            t.delno()
        else:
            print("Invalid choice!")

        ans = input("Do you want to continue? (1 for yes / 0 for no): ")
        if ans != '1':
            break

if __name__ == "__main__":
    main()

File: hashwithreplacement.cpp
#include <iostream>
#include <string.h>
#define max 10
using namespace std;
struct node
{
    char name[10];
    long int mn;
    int chain;
    node()
    {
        mn = 0;
        chain = -1;
        strcpy(name, "-----");
    }
};
class telephone
{
public:
    node ht[max];
    int count;
    telephone() { count = 0; }

    int hashfun(long int);
    void insert();
    void find();
    void delno();
    void display();
};

int telephone::hashfun(long int num)
{
    return (num % 10);
}

void telephone::insert()
{
    node s;
    int ind;
    cout << "Enter name and telephone number: \n";
    cin >> s.name >> s.mn;
    ind = hashfun(s.mn);
    
    if (count == max - 1)
        cout << "Hash table full\n";
    if (ht[ind].mn == 0)
    {
        ht[ind] = s;
        count++;
    }
    else if (hashfun(ht[ind].mn) == ind)
    {
        int prev = ind;
        while (ht[ind].mn != 0)
        {
            if (ht[ind].chain != -1){
                ind = ht[ind].chain;
                prev = ind;
            }else
            {
                while (ht[ind].mn != 0)
                    ind = (ind + 1) % max;
                ht[prev].chain = ind;
            }
        }
        ht[ind] = s;
        count++;
    }
    else
    {
        int newIndex = ind;
        while (ht[newIndex].mn != 0)
            newIndex = (newIndex + 1) % max;

        node temp = ht[ind];
        ht[newIndex] = temp;
        ht[ind] = s;

        int tempIndex = hashfun(temp.mn);
        while (ht[tempIndex].chain != ind)
            tempIndex = ht[tempIndex].chain;
        ht[tempIndex].chain = newIndex;

        ht[ind].chain = -1;
        count++;
    }
}

void telephone::find()
{
    long int num;
    int cnt = 0;
    cout << "Enter the telephone number to search: ";
    cin >> num;
    int ind = hashfun(num);

    while (ind != -1)
    {
        cnt++;
        if (ht[ind].mn == num)
        {
            cout << "Record Found!\n";
            cout << "Name: " << ht[ind].name << "\tPhone: " << ht[ind].mn << " found at index " << ind << " comparison " << cnt << endl;

            return;
        }
        ind = ht[ind].chain;
    }
    cout << "Record Not Found!\n";
}

void telephone::delno()
{
    long int num;
    cout << "Enter the telephone number to delete: ";
    cin >> num;
    int ind = hashfun(num);
    int prev = -1;

    while (ind != -1)
    {
        if (ht[ind].mn == num)
        {
            cout << "Record Deleted!\n";
            ht[ind].mn = 0;
            strcpy(ht[ind].name, "-----");
            if (prev != -1)
                ht[prev].chain = ht[ind].chain;
            ht[ind].chain = -1;
            return;
        }
        prev = ind;
        ind = ht[ind].chain;
    }
    cout << "Record Not Found!\n";
}

void telephone::display()
{
    for (int i = 0; i < max; i++)
    {
        cout << i << "\t" << ht[i].name << "\t" << ht[i].mn << "\t" << ht[i].chain << endl;
    }
}

int main()
{
    telephone t;
    int ch;
    char ans;
    do
    {
        cout << "Enter your choice" << endl;
        cout << "1.Insert" << endl;
        cout << "2.Display" << endl;
        cout << "3.Find" << endl;
        cout << "4.Delete" << endl;
        cin >> ch;
        switch (ch)
        {
        case 1:
            t.insert();
            break;
        case 2:
            t.display();
            break;
        case 3:
            t.find();
            break;
        case 4:
            t.delno();
            break;
        default:
            cout << "Invalid choice!!";
        }
        cout << "Do you want to continue? (1/0)";
        cin >> ans;
    } while (ans == '1');
    return 0;
}
